File: frontend/src/api/support.ts
import { apiClient } from './client'

export interface TicketEntry {
  id: string
  subject: string
  status: string
  priority: string
  category: string
  entity_type: string
  entity_id: string | null
  reporter_id: string | null
  reporter_name: string | null
  assigned_to: string | null
  assignee_name: string | null
  sla_deadline: string | null
  message_count: number
  created_at: string
  updated_at: string
  resolved_at: string | null
  closed_at: string | null
}

export interface TicketMessage {
  id: string
  sender_id: string | null
  sender_name: string | null
  body: string
  is_internal: boolean
  created_at: string
}

export interface TicketDetail extends TicketEntry {
  body: string | null
  messages: TicketMessage[]
  attachments: unknown[]
}

export interface TicketListResponse {
  total: number
  items: TicketEntry[]
}

export interface CreateTicketPayload {
  subject: string
  body?: string
  priority?: string
  category?: string
  context_job_id?: string
  context_application_id?: string
}

export interface AddMessagePayload {
  body: string
}

// ── Employer support API ──────────────────────────────────────────────────────

export const employerSupportApi = {
  createTicket: (data: CreateTicketPayload) =>
    apiClient.post<TicketEntry>('/employer/support/tickets', data),

  listTickets: () =>
    apiClient.get<TicketListResponse>('/employer/support/tickets'),

  getTicket: (ticketId: string) =>
    apiClient.get<TicketDetail>(`/employer/support/tickets/${ticketId}`),

  addMessage: (ticketId: string, data: AddMessagePayload) =>
    apiClient.post<TicketMessage>(`/employer/support/tickets/${ticketId}/messages`, data),
}

// ── Candidate support API ─────────────────────────────────────────────────────

export const candidateSupportApi = {
  createTicket: (data: CreateTicketPayload) =>
    apiClient.post<TicketEntry>('/me/support/tickets', data),

  listTickets: () =>
    apiClient.get<TicketListResponse>('/me/support/tickets'),

  getTicket: (ticketId: string) =>
    apiClient.get<TicketDetail>(`/me/support/tickets/${ticketId}`),

  addMessage: (ticketId: string, data: AddMessagePayload) =>
    apiClient.post<TicketMessage>(`/me/support/tickets/${ticketId}/messages`, data),
}

File: backend/app/models/support.py
"""Support ticket models for the admin support module."""
import uuid

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base


def _uuid():
    return uuid.uuid4()


TICKET_STATUSES = ("open", "pending", "resolved", "closed")
TICKET_PRIORITIES = ("low", "normal", "high", "urgent")
ENTITY_TYPES = ("employer", "candidate", "general")


class SupportTicket(Base):
    """A support ticket raised by or on behalf of a user."""
    __tablename__ = "support_tickets"

    id            = Column(UUID(as_uuid=True), primary_key=True, default=_uuid)
    subject       = Column(String(300), nullable=False)
    body          = Column(Text, nullable=True)  # initial description
    status        = Column(String(20), nullable=False, default="open", server_default="open",
                           index=True)
    priority      = Column(String(10), nullable=False, default="normal", server_default="normal",
                           index=True)

    # Who the ticket is about
    entity_type   = Column(String(20), nullable=False, default="general", server_default="general")
    entity_id     = Column(UUID(as_uuid=True), nullable=True)  # employer/aspirant user_id

    # Reporter — the user who submitted the ticket (may be null for admin-created)
    reporter_id   = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"),
                            nullable=True, index=True)

    # Assigned admin
    assigned_to   = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"),
                            nullable=True, index=True)

    # Category & optional context links
    category      = Column(String(40), nullable=False, default="general", server_default="general")
    context_job_id         = Column(UUID(as_uuid=True), ForeignKey("job_postings.id",    ondelete="SET NULL"), nullable=True)
    context_application_id = Column(UUID(as_uuid=True), ForeignKey("applications.id",   ondelete="SET NULL"), nullable=True)

    # SLA deadline (set on creation based on priority)
    sla_deadline  = Column(DateTime(timezone=True), nullable=True)

    created_at    = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)
    updated_at    = Column(DateTime(timezone=True), server_default=func.now(),
                            onupdate=func.now(), nullable=False)
    resolved_at   = Column(DateTime(timezone=True), nullable=True)
    closed_at     = Column(DateTime(timezone=True), nullable=True)

    __table_args__ = (
        CheckConstraint(f"status IN {TICKET_STATUSES}", name="ck_ticket_status"),
        CheckConstraint(f"priority IN {TICKET_PRIORITIES}", name="ck_ticket_priority"),
        CheckConstraint(f"entity_type IN {ENTITY_TYPES}", name="ck_ticket_entity_type"),
    )

    reporter    = relationship("User", foreign_keys=[reporter_id])
    assignee    = relationship("User", foreign_keys=[assigned_to])
    messages    = relationship("TicketMessage", back_populates="ticket",
                               cascade="all, delete-orphan", order_by="TicketMessage.created_at")
    attachments = relationship("TicketAttachment", back_populates="ticket",
                                cascade="all, delete-orphan")


class TicketMessage(Base):
    """A message in a support ticket thread. is_internal = admin-only note."""
    __tablename__ = "ticket_messages"

    id         = Column(UUID(as_uuid=True), primary_key=True, default=_uuid)
    ticket_id  = Column(UUID(as_uuid=True), ForeignKey("support_tickets.id", ondelete="CASCADE"),
                         nullable=False, index=True)
    sender_id  = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"),
                         nullable=True)
    body       = Column(Text, nullable=False)
    is_internal = Column(Boolean, nullable=False, default=False, server_default="false")
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    ticket  = relationship("SupportTicket", back_populates="messages")
    sender  = relationship("User", foreign_keys=[sender_id])


class TicketAttachment(Base):
    """A file attached to a support ticket."""
    __tablename__ = "ticket_attachments"

    id            = Column(UUID(as_uuid=True), primary_key=True, default=_uuid)
    ticket_id     = Column(UUID(as_uuid=True), ForeignKey("support_tickets.id", ondelete="CASCADE"),
                            nullable=False, index=True)
    uploaded_by   = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"),
                            nullable=True)
    file_key      = Column(String(500), nullable=False)   # S3 / storage key
    filename      = Column(String(300), nullable=False)
    content_type  = Column(String(100), nullable=True)
    size_bytes    = Column(Integer, nullable=True)
    created_at    = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

    ticket   = relationship("SupportTicket", back_populates="attachments")
    uploader = relationship("User", foreign_keys=[uploaded_by])
